eessi/testsuite/tests/apps/openfoam/openfoam.py

```diff
@@ -0,0 +1,150 @@
+
+"concrete test cases": all test combinations that will actually run:
+- test variants
+- valid system:partition+programming environment combinations
+https://reframe-hpc.readthedocs.io/en/stable/tutorial_deps.html#listing-dependencies
+
+Tests can be filtered by name, tag, programming environment, system, partition, or maintainer,
+see https://reframe-hpc.readthedocs.io/en/stable/manpage.html#test-filtering
+
+Hooks acting on all possible test combinations (before filtering) are called after the 'init' stage.
+Hooks acting on concrete test cases (after filtering) are called after the 'setup' stage.
+
+See also https://reframe-hpc.readthedocs.io/en/stable/pipeline.html
+"""
+
+import reframe as rfm
+# added only to make the linter happy
+from reframe.core.builtins import deferrable, parameter, run_after, run_before, sanity_function, performance_function
+import reframe.utility.sanity as sn
+from eessi.testsuite.constants import COMPUTE_UNITS, DEVICE_TYPES, SCALES
+from eessi.testsuite.eessi_mixin import EESSI_Mixin
+from eessi.testsuite.utils import find_modules
+
+
+def filter_scales_8M():
+    """
+    Filtering function for filtering scales for the OpenFOAM 8M mesh test
+    returns all scales with at least 2 nodes.
+    """
+    return [
+        k for (k, v) in SCALES.items()
+        if (v['num_nodes'] > 1) and (v.get('node_part', 0) == 1)
+    ]
+
+
+@rfm.simple_test
+class EESSI_OPENFOAM_LID_DRIVEN_CAVITY(rfm.RunOnlyRegressionTest, EESSI_Mixin):
+    """
+    This is the main OPENFOAM class for the Lid-driven cavity test. The test consists of many steps which are run as
+    pre-run commands and the main test with the executable `icoFoam` is measured for performance.
+    """
+#    ldc_8M = fixture(EESSI_OPENFOAM_base, scope='partition')
+    executable = 'icoFoam'
+    executable_opts = ['-parallel', '2>&1', '|', 'tee log.icofoam']
+    time_limit = '60m'
+    readonly_files = ['']
+    device_type = parameter([DEVICE_TYPES.CPU])
+    module_name = parameter(find_modules('OpenFOAM/v', name_only=False))
+    valid_systems = ['*']
+    scale = parameter(filter_scales_8M())
+
+    @run_after('init')
+    def set_compute_unit(self):
+        """
+        Set the compute unit to which tasks will be assigned:
+        one task per CPU core for CPU runs.
+        """
+        if self.device_type == DEVICE_TYPES.CPU:
+            self.compute_unit = COMPUTE_UNITS.CPU
+        else:
+            msg = f"No mapping of device type {self.device_type} to a COMPUTE_UNITS was specified in this test"
+            raise NotImplementedError(msg)
+
+    def required_mem_per_node(self):
+        return self.num_tasks_per_node * 1700
+
+    @run_after('setup')
+    def check_launcher_options(self):
+        if (self.job.launcher.command(self.job)[0] == 'mpirun'):
+            self.launcher_command = self.job.launcher.command(self.job)
+            self.launcher_command[-1] = str(self.num_tasks_per_node * self.num_nodes)
```

Comment:
@satishskamath Can you add some comments here to explain why this is necessary when `mpirun` is used as launcher, since we don't need to do this at all for other tests?

Reply:
I agree, a comment here to state why this is needed will be helpful in the future. I'd make sure it states two things:
1. a typical OpenFOAM job would contain multiple mpirun's
2. We had trouble using ReFrame fixtures to separate these over multiple jobs, because we couldn't get it to work together with the mixin class.